### ISSUE TITLE
Potential issue with python / event-node: “Task was destroyed but it is pending!”

### ISSUE THREAD

**Issue by skofgar** [Accenture/mercury-python#5]
# Situation

I'm trying to build a small application that is gathering communication from the network stack using [scapy](https://scapy.net). Scapy generates a lot of data that I want to send to a different computer via PostOffice.

In an initial attempt I just want to send the data to a microservices, which works fine for a bit, but if you leave the script running for about 30 seconds and browse the web, it will eventually start throwing error messages. 

Then it will no longer be possible to exit the application using `CTRL+C`

# Reproduce

1. start `language-connector-1.12.9.jar`
2. start `event-node-1.12.9.jar`
3. install `scapy` and make sure to use python `3.6.7` (if you use [conda](https://www.anaconda.com) to configure the environment with `conda env create -f environment.yml`, make sure to then activate it: `conda activate sniffer`)
4. modify `sniffer-error.py`: set `COMPUTER_INTERFACE` to your active internet interface (use `ifconfig`)
5. run `sniffer-error.py` using `python sniffer-error.py`
6. after about 30 seconds of running and internet usage, errors should occur and `CTRL+C` should no longer allow aborting the program

```
ERROR base_events.py:1274 Task exception was never retrieved
future: <Task finished coro=<ServiceQueue.listen() done, defined at /Repo/my-app/src/mercury/mercury/platform.py:94> exception=TypeError('Cannot serialize <Ether...
```

# My resolution 

> TL;DR: none

My research so far lead me to believe that there might be an issue with the _asyncio_ loop, that it gets stopped or exceptions are not properly handled. I did try to add exception handling to the [WorkerQueue](https://github.com/Accenture/mercury-python/blob/master/mercury/platform.py#L172), but without success. 

I will investigate this issue further, but thought I would share it in the meantime.

Potential leads:
- [StackOverflow - “Task was destroyed but it is pending!”](https://stackoverflow.com/questions/40897428/please-explain-task-was-destroyed-but-it-is-pending)

[sniffer.zip](https://github.com/Accenture/mercury-python/files/3640586/sniffer.zip)

**Comment by ericcwlaw** [comment]
this seems to be a serialization problem. The underlying serialization engine is MsgPack. The object to be serialized should be a dictionary of basic types. It cannot serialize proprietary objects.

**Comment by skofgar** [comment]
That sounds like the problem! 

Is there a way I can set the payload to binary? I'm looking at the [`isBinary` sections of the mercury core](https://github.com/Accenture/mercury/blob/894e6ab3e5f0273eca0f73dd1c270699561a88a7/system/platform-core/src/main/java/org/platformlambda/core/models/EventEnvelope.java#L331). Converting it into dictionaries of basic types is probably the better way to go.

**Comment by skofgar** [comment]
I also tested, simply sending a dictionary, e.g. replacing the `activateSniffer` function with:

```
def activateSniffer(headers: dict, body: any):
    #### this loop causes ClassNotFoundException, IllegalArgumentException in the event-node
    for i in range(1,1_000):
         forwardPacket({"value": i})
```
for convenience, the `forwardPacket` method basically just calls `po.send(...)`. 
```
def forwardPacket(packet):
    Platform().log.info("package type " + str(type(packet)))
    po.send("server.analyze.packet", headers={}, body=packet)
```

However, this causes `ClassNotFoundException` and `IllegalArgumentException` in the event-node. Review [event-node-1.12.9-excerpt.log](https://github.com/Accenture/mercury-python/files/3640848/event-node-1.12.9-excerpt.log)

But this is likely a different error.

**Comment by ericcwlaw** [comment]
In Python 3, str and bytes are no longer the same. You can directly send binary payload. We will investigate the jax.rs issue.

**Comment by ericcwlaw** [comment]
the isBinary in Java side allows the serialization to fall back to JSON string for some edge cases. It should not be used with Python. The protocols are incompatible.

**Comment by skofgar** [comment]
I can confirm, that switching to a string, for example, solves the problem, or using pickle, but after talking with @ericcwlaw writing a custom wrapper looks like a better approach.
